mlflow/utils/model_utils.py

```diff
@@ -428,3 +429,70 @@ def _validate_pyfunc_model_config(model_config):
             "valid file path or of type ``dict`` with string keys.",
             error_code=INVALID_PARAMETER_VALUE,
         )
+
+
+RECORD_ENV_VARS = {
+    # api key related
+    "API_KEY",
+    # databricks auth related
+    "DATABRICKS_HOST",
+    "DATABRICKS_USERNAME",
+    "DATABRICKS_PASSWORD",
+    "DATABRICKS_TOKEN",
+    "DATABRICKS_INSECURE",
+    "DATABRICKS_CLIENT_ID",
+    "DATABRICKS_CLIENT_SECRET",
+    "_DATABRICKS_WORKSPACE_HOST",
+    "_DATABRICKS_WORKSPACE_ID",
+    # dbconnect related
+    "SPARK_REMOTE",
+    "SPARK_LOCAL_REMOTE",
+}
+
+
+@contextlib.contextmanager
+def env_var_tracker():
+    """
+    Context manager for temporarily tracking environment variables accessed.
+    It tracks both environment variables set and accessed during the context manager's lifetime.
+    """
+    from mlflow.environment_variables import MLFLOW_RECORD_ENV_VARS_IN_MODEL_LOGGING
+
+    if MLFLOW_RECORD_ENV_VARS_IN_MODEL_LOGGING.get():
+        original_setitem = os._Environ.__setitem__
+        original_getitem = os._Environ.__getitem__
+        original_get = os._Environ.get
+        tracked_env_names = set()
+
+        def updated_set_item(self, key, value):
+            original_setitem(self, key, value)
+            if any(env_var in key for env_var in RECORD_ENV_VARS):
+                tracked_env_names.add(key)
+
+        def updated_get_item(self, key):
+            result = original_getitem(self, key)
+            if any(env_var in key for env_var in RECORD_ENV_VARS):
+                tracked_env_names.add(key)
+            return result
+
+        def updated_get(self, key, *args, **kwargs):
+            if key in self and any(env_var in key for env_var in RECORD_ENV_VARS):
+                tracked_env_names.add(key)
+            return original_get(self, key, *args, **kwargs)
+
+        def get_tracked_env_names(self):
+            return tracked_env_names
+
+        try:
+            os._Environ.__getitem__ = updated_get_item
+            os._Environ.__setitem__ = updated_set_item
```

Comment:
Wondering if we need to check `_setitem_`? This means the env var is set but never used🤔

Reply:
This is a good point :) I think we can drop _setitem_ track!